internal/status/conditions_test.go

```diff
@@ -49,3 +53,183 @@ func TestHTTPRouteAddCondition(t *testing.T) {
+				Status:             metav1.ConditionTrue,
+				LastTransitionTime: metav1.Unix(1, 0),
+			},
+		},
+		{
+			name:     "check condition reason differs",
+			expected: true,
+			a: metav1.Condition{
+				Type:   string(gatewayapi_v1alpha1.GatewayConditionReady),
+				Status: metav1.ConditionFalse,
+				Reason: "foo",
+			},
+			b: metav1.Condition{
+				Type:   string(gatewayapi_v1alpha1.GatewayConditionReady),
+				Status: metav1.ConditionFalse,
+				Reason: "bar",
+			},
+		},
+		{
+			name:     "condition status differs",
+			expected: true,
+			a: metav1.Condition{
+				Type:   string(gatewayapi_v1alpha1.GatewayClassConditionStatusAdmitted),
+				Status: metav1.ConditionTrue,
+			},
+			b: metav1.Condition{
+				Type:   string(gatewayapi_v1alpha1.GatewayClassConditionStatusAdmitted),
+				Status: metav1.ConditionFalse,
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		if actual := conditionChanged(tc.a, tc.b); actual != tc.expected {
+			t.Fatalf("%q: expected %v, got %v", tc.name, tc.expected, actual)
+		}
+	}
+}
+
+func TestMergeConditions(t *testing.T) {
+	// Inject a fake clock and don't forget to reset it
+	fakeClock := utilclock.NewFakeClock(time.Time{})
+	clock = fakeClock
+	defer func() {
+		clock = utilclock.RealClock{}
+	}()
+
+	start := fakeClock.Now()
+	middle := start.Add(1 * time.Minute)
+	later := start.Add(2 * time.Minute)
+
+	testCases := []struct {
+		name     string
+		current  []metav1.Condition
+		updates  []metav1.Condition
+		expected []metav1.Condition
+	}{
+		{
+			name: "status updated",
+			current: []metav1.Condition{
+				newCondition("available", "false", "Reason", "Message", start),
+			},
+			updates: []metav1.Condition{
+				newCondition("available", "true", "Reason", "Message", middle),
+			},
+			expected: []metav1.Condition{
+				newCondition("available", "true", "Reason", "Message", later),
+			},
+		},
+		{
+			name: "reason updated",
+			current: []metav1.Condition{
+				newCondition("available", "false", "Reason", "Message", start),
+			},
+			updates: []metav1.Condition{
+				newCondition("available", "false", "New Reason", "Message", middle),
+			},
+			expected: []metav1.Condition{
+				newCondition("available", "false", "New Reason", "Message", start),
+			},
+		},
+		{
+			name: "message updated",
+			current: []metav1.Condition{
+				newCondition("available", "false", "Reason", "Message", start),
+			},
+			updates: []metav1.Condition{
+				newCondition("available", "false", "Reason", "New Message", middle),
+			},
+			expected: []metav1.Condition{
+				newCondition("available", "false", "Reason", "New Message", start),
+			},
+		},
+	}
+
+	// Simulate the passage of time between original condition creation
+	// and update processing
+	fakeClock.SetTime(later)
+
+	for _, tc := range testCases {
+		actual := mergeConditions(tc.current, tc.updates...)
+		if conditionChanged(tc.expected[0], actual[0]) {
+			t.Errorf("expected:\n%v\nactual:\n%v", tc.expected, actual)
```

Comment:
We've used testify `assert.Equal(t, want, got)` in other places, might simplify the error message.